<amp-apester-media> fix random embed and improve performance

OmriKeret opened the pull request:
Hi Guys, 
This PR is is ment for fixing broken embedding option for the amp-apester-media component and some performance improvment.

Summary
- Fix embedding of a random media.
- Feature - added Apester feed support.
- Preconnect to our static resources server in order to speed up resource fetching.

Thanks,
  Omri Keret @ Apester

OmriKeret commented:
We already signed on the CLA @mrsufgi

aghassemi commented:
@OmriKeret Thansk for the PR. Regarding CL, I see that Apester has signed a corp CLA, however CLA check is not passing. Could you please make sure your email (oria@apester.com) is a member of the Group that was used to sign the corp CLA?

OmriKeret commented:
My Email should be added to the group today (omri.keret@gmail.com).
I will renotify after it's been added @aghassemi 

Thanks!

aghassemi commented on extensions/amp-apester-media/0.1/amp-apester-media.css:
```diff
@@ -122,3 +123,13 @@
+  z-index:    2147483646 !important;
+  zoom:       1;
+  -webkit-overflow-scrolling: touch !important;
+}
```
nit: new line after this line

aghassemi commented on extensions/amp-apester-media/0.1/amp-apester-media.js:
```diff
@@ -18,29 +18,45 @@ import {user, dev} from '../../../src/log';
 /** @const */
 const TAG = 'amp-apester-media';
+/**
+ * @type {{SET_FULL_SCREEN: string, REMOVE_FULL_SCREEN: string, RESIZE_UNIT: string}}
```
`@enum {string}` instead

aghassemi commented on extensions/amp-apester-media/0.1/amp-apester-media.js:
```diff
@@ -63,6 +79,16 @@ class AmpApesterMedia extends AMP.BaseElement {
      */
     this.mediaAttribute_ = null;
+    /**
+     * @private {?string}
```
type does not match  `{!Object}` instead ?

aghassemi commented on extensions/amp-apester-media/0.1/amp-apester-media.js:
```diff
@@ -101,31 +128,62 @@ class AmpApesterMedia extends AMP.BaseElement {
-        `/interactions/${encodedMediaAttribute}/display`;
-    return `${this.displayBaseUrl_}${suffix}`;
+    let suffix = '';
+    let queryParmas = '?';
```
please turn this into a map to store key/values and then use `addParamToUrl` util from `url.js` to append to the url. the util will handle encoding as well.

aghassemi commented on extensions/amp-apester-media/0.1/amp-apester-media.js:
```diff
@@ -142,18 +200,41 @@ class AmpApesterMedia extends AMP.BaseElement {
    * */
   constructUrlFromMedia_(id) {
-    return `${this.rendererBaseUrl_}/interaction/${encodeURIComponent(id)}`;
+    const conditionalChannelId = this.embedOptions_distributionChannelId ?
```
ditto: use provided utils from `url.js`

aghassemi commented on extensions/amp-apester-media/0.1/amp-apester-media.js:
```diff
@@ -142,18 +200,41 @@ class AmpApesterMedia extends AMP.BaseElement {
+      `&type=${encodeURIComponent(this.embedOptions_.playlist ? 'playlist' :
+        'editorial')}`;
+    const platform = `&platform=${getPlatform()}`;
+    const cannonicalUrl = document.querySelector('link[rel=\'canonical\']')
```
use `Services.documentInfoForDoc(this.element).canonicalUrl` to get the Canonical Url.

aghassemi commented on extensions/amp-apester-media/0.1/amp-apester-media.js:
```diff
@@ -142,18 +200,41 @@ class AmpApesterMedia extends AMP.BaseElement {
     iframe.setAttribute('scrolling', 'no');
-    iframe.src = src;
+    iframe.setAttribute('data-interaction-id', media.interactionId);
+    iframe.name = /wpcomwidgets/.test(document.location.href) ?
```
can you explain what's happening here? the href would be the AMP page using the component and in post cases an AMP CDN Cahe URL, what does the check for `wpcomwidgets` supposed to do? Also use `this.win.location.href` instead of `document.location`

OmriKeret replied:
Well, we sometimes get embedded in a WP.com widget iframe. In that case we are using the referrer in order to understand where is the current media unit lives. 
But, after rethinking I guess that it won't be happening with an AMP page so I will remove it.

aghassemi commented on extensions/amp-apester-media/0.1/amp-apester-media.js:
```diff
@@ -142,18 +200,41 @@ class AmpApesterMedia extends AMP.BaseElement {
+    iframe.name = /wpcomwidgets/.test(document.location.href) ?
+      document.referrer : document.location.href;
+    const splittedUrl = src.split('?');
+    iframe.originalUrlParams = splittedUrl.length > 1 ? `?${splittedUrl[1]}`
```
confused here, your embedded page can't access these values (it can only access `name`), so why setting them?

aghassemi commented on extensions/amp-apester-media/0.1/amp-apester-media.js:
```diff
@@ -244,23 +329,30 @@ class AmpApesterMedia extends AMP.BaseElement {
+
+          return this.iframePromise_ =
+        this.getMediaContent(media.content, src).then(content => {
+          return installFriendlyIframeEmbed(this.iframe_, this.element, {
```
we can't allow your dynamic content to be in a friendly iframe which has access to the top-level page context. Only reviewed code can access top-level context. /cc @dvoytenko to confirm my statement.

OmriKeret replied:
Ho.. We are getting back the HTML for the media back from our servers.
No way to leverage it on the AMP pages?

aghassemi replied:
Not outside of a 3P iframe. We can't allow your HTML access to the 1P page (FriendlyIframe is same origin and has access to the page). 

What you like to do is certainly possible but you would need to use our 3P frame approach (amp-facebook, amp-twitter, etc.. do this already). In this approach, we will create an iframe with a different origin for you and your specified JS file runs inside that frame, it can fetch whatever and replace the content as needed. Communication between the component and the frame still need to be through post messages.

aghassemi commented on extensions/amp-apester-media/0.1/utils.js:
```diff
@@ -0,0 +1,105 @@
+ * @return {Array} array of keywords
+ */
+function extractMetaTags(doc) {
+  const selectors = ['meta[name=keywords]', 'meta[name=news_keywords]',
```
/cc @rudygalfi @dvoytenko are extensions allowed to read and pass this data?

OmriKeret replied:
We are using this data in order to get contextual content for the required page.
The content is picked from the publisher contents pool.

aghassemi replied:
Got it. but AMP Cache removes these metadata to optimize the size of the page so in majority of cases you won't find them anyway. 

If this is a critical feature, please create an issue so the team can assess: 1- have cache not remove them 2- decide whether it is fine to allow 3Ps to pass this data to themselves.

aghassemi commented:
@OmriKeret Friendly ping

OmriKeret commented:
@aghassemi  Sorry for the long break.
I've been informed today that my Email supposed to be in the contributor list now, so we could re run the cla check? 

Other than that, I've fixed your comments (removed the friendly Iframe embed) and fixed linting issues.

Could you please review again?
Thanks :)

aghassemi commented:
@OmriKeret there are bunch of lint errors and presubmit / unit tests check is red, please fix and I will do the final review when all are green.

For CLA, please reply here with `I Signed It` and CLA bot should check again.

Thanks!

OmriKeret commented:
I Signed It

OmriKeret commented:
Hi @aghassemi, I think I fixed all of the linting issues.
I don't know why the gulp visual-diff fails on travis. Locally I managed to run it with no errors.
Can you shed some light on the problem?

aghassemi commented:
@OmriKeret could you please rebase the PR? That should fix the visual test. I will take a look soon. Thanks for your patience.

OmriKeret commented:
Hi @aghassemi, I rebased and now I fail on the integration tests on Sauce Labs.
Not sure if it's something I've caused.

aghassemi commented on extensions/amp-apester-media/0.1/amp-apester-media.js:
```diff
@@ -292,14 +391,75 @@ class AmpApesterMedia extends AMP.BaseElement {
+   * @param content
+   * @param src
+   */
+  getMediaContent(content, src) {
```
not used anymore, please remove

OmriKeret replied:
Removed

aghassemi commented on extensions/amp-apester-media/0.1/utils.js:
```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright 2016 The AMP HTML Authors.
```
nit: 2017

OmriKeret replied:
oops, changed.

aghassemi commented on extensions/amp-apester-media/0.1/utils.js:
```diff
@@ -0,0 +1,101 @@
+  // if it says it's a webview, let's go with that
+  'WebView',
+  // iOS webview will be the same as safari but missing "Safari"
+  '(iPhone|iPod|iPad)(?!.*Safari)',
```
"Standalone" apps (e.g. web-app enabled and added to homescreen) also miss "Safari" so they may not be very accurate.

We ask webview AMP viewers to tell the AMP page, that's exposed under `viewer.isWebviewEmbedded()` from viewer. You may want to use that instead. Up to you, not sure how much accuracy of this matters to you.

OmriKeret replied:
The accuracy is enough for our use. So we will keep it that way for now :)

OmriKeret commented:
Hi @aghassemi, thanks for the review.
I made the required changes, I hope we are good to go now :)

aghassemi commented:
@OmriKeret LGTM, CLA is still an issue though. The commits on this PR are from `oria@apester.com` but `omri.keret@gmail.com` is the one added to the Corp CLA group for `apester`. Could you please add  `oria@apester.com` to the same Corp CLA group and reply "I signed it!"?

OmriKeret commented:
Thanks @aghassemi! oria@apester.com should already be in the CLA group.
He had already contributed some code to AMP for Apester.
Maybe it's me that wasn't added yet? 
I will make a recheck anyway.

Thanks!

aghassemi commented:
@OmriKeret Yeah maybe omri.keret@gmail.com is not added yet. There are commits from both omri.keret@gmail.com and oria@apester.com  so both need to be in the group.

OmriKeret commented:
I signed it!

OmriKeret commented:
I signed it!

OmriKeret commented:
HI @aghassemi, I've been told I was added to the group.
But the Bot didn't yet approved me. Can you help out here?

Thanks!

mrjoro commented:
I looked into the CLA issue and the original problem was that the Google Group that's associated with the corp CLA was deleted at some point.  A new group has been created and associated with the CLA, so hopefully once the owner of that group (at Apester) adds the email addresses to it the CLA bot should be happy.  Sometimes you have to prod the CLA bot (which I've tried to do by closing and reopening the issue) but if you are sure the email on the commit is in that group we can do a manual check if needed.

Sorry for the issues this is causing!

aghassemi commented:
Thanks @mrjoro for the CLA investigation.
@OmriKeret PR looks great. Thanks for improving your AMP component and hence the AMP ecosystem. Merging the PR now.
